src/libraries/System.Linq/src/System/Linq/Chunk.cs

```diff
@@ -45,15 +45,31 @@ public static IEnumerable<TSource[]> Chunk<TSource>(this IEnumerable<TSource> so
                 ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.size);
             }
 
-            if (IsEmptyArray(source))
+            if (source is TSource[] array)
             {
-                return [];
+                // Special-case arrays, which have an immutable length. This enables us to not only do an
```

Comment:
Couldn't we extend that assumption to more types if this method was returning an enumerator instead of an enumerable? IIRC we're already doing this in a few places, notably `Take(Range)`.

Reply:
Not following. You mean implement a custom enumerable to type and length test in GetEnumerator? And if the length changes while enumerating, do what?

Reply:
Correct.

> You mean implement a custom enumerable to type and length test in GetEnumerator? And if the length changes while enumerating, do what?

Undefined behavior, presumably.  Currently it should throw an exception because of the version checks, but we are considering removing those altogether.